java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java

```diff
@@ -167,8 +136,35 @@ private boolean isTestMethod(Symbol method) {
     });
   }
 
+  private boolean isTestAnnotation(Type type) {
+    return testAnnotations.contains(type.fullyQualifiedName()) || isJUnitTestableMetaAnnotated(type);
+  }
+
+  private boolean isJUnitTestableMetaAnnotated(Type type) {
+    if (seenAnnotations.contains(type.fullyQualifiedName())) {
+      return false;
+    }
+    seenAnnotations.add(type.fullyQualifiedName());
+    SymbolMetadata metadata = type.symbol().metadata();
+    if (metadata.isAnnotatedWith("org.junit.platform.commons.annotation.Testable")) {
+      testAnnotations.add(type.fullyQualifiedName());
+      return true;
+    }
+    for (SymbolMetadata.AnnotationInstance annotation : metadata.annotations()) {
+      if (isJUnitTestableMetaAnnotated(annotation.symbol().type())) {
+        testAnnotations.add(type.fullyQualifiedName());
+        return true;
+      }
+    }
+    return false;
+  }
+
   private static Stream<Symbol.MethodSymbol> getAllMembers(Symbol.TypeSymbol symbol, boolean isEnclosed) {
-    if ("java.lang.Object".equals(symbol.type().fullyQualifiedName())) {
+    return getAllMembers(symbol, isEnclosed, new HashSet<>());
+  }
+
+  private static Stream<Symbol.MethodSymbol> getAllMembers(Symbol.TypeSymbol symbol, boolean isEnclosed, Set<Symbol> visitedSymbols) {
+    if (!visitedSymbols.add(symbol) || "java.lang.Object".equals(symbol.type().fullyQualifiedName())) {
```

Comment:
Minor improvement: you could use
```suggestion
    if (!visitedSymbols.add(symbol) || symbol.type().is("java.lang.Object")) {
```
instead